module/zfs/arc.c

```diff
@@ -3995,6 +3994,15 @@ arc_evict_state_impl(multilist_t *ml, int idx, arc_buf_hdr_t *marker,
 
 	multilist_sublist_unlock(mls);
 
+	/*
+	 * If the ARC size is reduced from arc_c_max to arc_c_min (especially
+	 * if the average cached block is small), eviction can be on-CPU for
+	 * many seconds.  To ensure that other threads that may be bound to
+	 * this CPU are able to make progress, make a voluntary preemption
+	 * call here.
+	 */
+	cond_resched();
```

Comment:
It's a little surprising that we don't already get scheduled out fair frequently.  We must rarely take the slow path when acquiring the multilist mutexes due to lack of contention and optimistic spinning when acquiring.

Reply:
Yeah I also find it surprising.  We don't quite have airtight evidence for this (@grwilson is working on it), but it seems like the bio worker threads are getting starved out by this thread, causing zio "delay" reports.  These reports are always accompanied by ARC shrinking (to arc_c_min) events.